src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs

```diff
@@ -1961,9 +1994,55 @@ private SqlDataReader EndExecuteReaderInternal(IAsyncResult asyncResult)
             }
         }
 
-        private IAsyncResult BeginExecuteReaderAsync(CommandBehavior behavior, AsyncCallback callback, object stateObject)
+        private void CleanupExecuteReaderAsync(Task<SqlDataReader> task, TaskCompletionSource<SqlDataReader> source, Guid operationId)
         {
-            return BeginExecuteReaderInternal(behavior, callback, stateObject, CommandTimeout, inRetry: false, asyncWrite: true);
+            if (task.IsFaulted)
+            {
+                Exception e = task.Exception.InnerException;
+                if (!_parentOperationStarted)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, _transaction, e);
+                }
+                source.SetException(e);
+            }
+            else
+            {
+                if (task.IsCanceled)
+                {
+                    source.SetCanceled();
+                }
+                else
+                {
+                    source.SetResult(task.Result);
+                }
+                if (!_parentOperationStarted)
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, _transaction);
+                }
+            }
+        }
+
+        private static IAsyncResult BeginExecuteReaderAsyncCallback(AsyncCallback callback, object stateObject)
+        {
+            ExecuteReaderAsyncCallContext args = (ExecuteReaderAsyncCallContext)stateObject;
+            return args.Command.BeginExecuteReaderInternal(args.CommandBehavior, callback, stateObject, args.Command.CommandTimeout, inRetry: false, asyncWrite: true);
+        }
+
+        private static SqlDataReader EndExecuteReaderAsyncCallback(IAsyncResult asyncResult)
+        {
+            ExecuteReaderAsyncCallContext args = (ExecuteReaderAsyncCallContext)asyncResult.AsyncState;
+            return args.Command.EndExecuteReaderAsync(asyncResult);
+        }
+
+        private static void CleanupExecuteReaderAsyncCallback(Task<SqlDataReader> task)
+        {
+            ExecuteReaderAsyncCallContext context = (ExecuteReaderAsyncCallContext)task.AsyncState;
+            SqlCommand command = context.Command;
+            Guid operationId = context.OperationID;
+            TaskCompletionSource<SqlDataReader> source = context.TaskCompletionSource;
+            context.Dispose();
+
```

Comment:
Why not call Dispose inside CleanupExecuteReaderAsync?

Reply:
Cleaning it up here returns it to the cache earlier. It also releases any resources being held by the context object (buffers etc).